src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemDatasetValidator.ChangeState.cs

```diff
@@ -0,0 +1,104 @@
+
+    protected override void OnValidate(DicomDataset dataset)
+    {
+        // Check for missing Transaction UID
+        dataset.ValidateRequirement(DicomTag.TransactionUID, RequirementCode.OneOne);
+
+        // Check for missing Procedure Step State
+        dataset.ValidateRequirement(DicomTag.ProcedureStepState, RequirementCode.OneOne);
+
+        // Check for allowed procedure step state value
+        var targetProcedureStepStateStringValue = dataset.GetString(DicomTag.ProcedureStepState);
+        var targetProcedureStepState = ProcedureStepStateExtensions.GetProcedureStepState(targetProcedureStepStateStringValue);
+        if (!AllowedTargetStatesForWorkitemChangeState.Contains(targetProcedureStepState))
+        {
+            throw new DatasetValidationException(
+                FailureReasonCodes.ValidationFailure,
+                string.Format(
+                    CultureInfo.InvariantCulture,
+                    DicomCoreResource.UnexpectedValue,
+                    DicomTag.ProcedureStepState,
+                    string.Join(@",", AllowedTargetStatesForWorkitemChangeState)));
+        }
+    }
+
+    /// <summary>
+    /// Validates Workitem state in the store and procedure step state transition validity.
+    ///
+    /// Throws <see cref="DatasetValidationException"/> when the workitem-metadata status is not read-write.
+    /// Throws <see cref="DatasetValidationException"/> when the workitem-metadata transition state has error.
+    /// 
+    /// </summary>
+    /// <param name="dataset">The Change Workitem State request DICOM dataset</param>
+    /// <param name="workitemMetadata">The Workitem Metadata</param>
+    internal static void ValidateWorkitemState(DicomDataset dataset, WorkitemMetadataStoreEntry workitemMetadata)
+    {
+        EnsureArg.IsNotNull(dataset, nameof(dataset));
+        EnsureArg.IsNotNull(workitemMetadata, nameof(workitemMetadata));
+
+        // the request Transaction UID must match the current Transaction UID.
+        var transactionUid = dataset.GetString(DicomTag.TransactionUID);
+        if (!string.IsNullOrWhiteSpace(workitemMetadata.TransactionUid) &&
+            !string.Equals(workitemMetadata.TransactionUid, transactionUid, System.StringComparison.Ordinal))
+        {
+            throw new DatasetValidationException(
+                FailureReasonCodes.ValidationFailure,
+                DicomCoreResource.InvalidTransactionUID);
+        }
+
+        // Check for the transition state rule validity
+        var targetProcedureStepStateStringValue = dataset.GetString(DicomTag.ProcedureStepState);
+        var targetProcedureStepState = ProcedureStepStateExtensions.GetProcedureStepState(targetProcedureStepStateStringValue);
+
+        var actionEvent = (targetProcedureStepState == ProcedureStepState.InProgress)
+            ? WorkitemStateEvents.NActionToInProgressWithCorrectTransactionUID
+            : (targetProcedureStepState == ProcedureStepState.Canceled)
```

Comment:
If the target state is Canceled, shouldn't we call our canceled method?

Reply:
Done